### PR TITLE
perf: Apply Kahan compensation to incremental Ewald structure factor accumulation

### PR DESCRIPTION
**Pull request by n-gao**
Kahan compensation is now applied to the structure factor accumulator (`S(k)`) used in incremental Monte Carlo updates. Previously, each accepted move added a floating-point delta directly to the running sum, causing small deltas to lose low-order bits against a large accumulated `S(k)` and producing drift over long move chains. The compensation term carries those dropped bits forward into subsequent updates.

Concretely:
- `EwaldCache.structure_factor` is now a `KahanSummand[Array]` instead of a bare `Array`, and its initializer wraps the zero array in `KahanSummand.init`.
- `EwaldCachePatch` stores and applies the full `KahanSummand` (value + compensation) on acceptance, so the compensation is not discarded between moves.
- `_structure_factor_update` returns a `KahanSummand` by delegating to the `+` operator on the cached accumulator.
- The JVP for `_structure_factor_update` returns a `KahanSummand` primal and carries a zero tangent for the compensation term, which is exactly zero in real arithmetic.
- `structure_factor` wraps full recomputations in `KahanSummand.init` (fresh compensation) and returns the accumulator; callers read `S(k)` via `.total`.
- `ewald_long_range_energy` passes `structure_out.total` to `long_range`.

A new test class `TestStructureFactorCache` verifies that carrying the compensation across 1000 single-particle moves on a charge-ordered cubic lattice (in single precision, where accumulation error is most visible) reduces the maximum absolute error in `S(k)` by more than 4× compared to dropping the compensation each step.

---

## Results

Heat of adsorption against a float64 reference, five charged frameworks labelled by mean loading:

| framework | mean loading | float64 (eV) | float32 before | float32 after |
|---|---|---|---|---|
| A | 251.7 | -0.44519 | -0.61295 (-37.7%) | -0.43612 (+2.0%) |
| B | 196.1 | -0.41887 | -0.49177 (-17.4%) | -0.42042 (-0.4%) |
| C | 10.2 | -0.34918 | -0.35486 (-1.6%) | -0.35663 (-2.1%) |
| D | 3.5 | -0.30465 | -0.30111 (+1.2%) | -0.30653 (-0.6%) |
| E | 9.4 | -0.31797 | -0.31956 (-0.5%) | -0.32221 (-1.3%) |

Both high-loading frameworks are now within replica error, the scatter on A falls from +-0.05329 to +-0.01188 eV (float64 +-0.01572), and the mean energy error drops from -13.98 to +0.35 eV on A and -18.95 to +0.27 eV on B. The cause was the position-independent self-interaction term being accumulated incrementally, so each per-move delta was the difference of two full-system sums (~710 eV on A), injecting ~1/3 ulp per accepted move and compounding to -3.2 eV over ~150k moves; recomputing it in full drops that term's drift to -1.5e-5 eV. Acceptance was never affected (~2e-5 eV, about 1e-3 kT per move), so uptake always agreed within 1.5 sigma. The float64 run is bit-identical before and after, and the suite is unchanged at 1895 passed / 7 skipped.

Setup: batched grand-canonical Monte Carlo of CO2, 328-1016 atoms per cell, 4 kPa, 313.15 K, 4 replicas per framework (20 systems), 300 warmup + 3000 production cycles, same seed in both precisions.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#269** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/269?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#278** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/278?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#272** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/272?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#271** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/271?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#268** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/268?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/268?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#267** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/267?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#229** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/229?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#233** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/233?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#235** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/235?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#234** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/234?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#232** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/232?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#230** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#228** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/228?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#227** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#226** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/226?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#224** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#223** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/223?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#222** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/222?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#215** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/215?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#213** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/213?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#216](https://github.com/cusp-ai-oss/kUPS/pull/216) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#212** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#211** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/211?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#210** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/210?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#204** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#209** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#203** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#202** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#201** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#186** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#185** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#183** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#182** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#181** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/181?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#180** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/180?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#178** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#206** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#205** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/kups/potential/classical/ewald.py [hunk 1]
```diff
@@ -105,14 +105,15 @@ class EwaldCache[Gradient, Hessian]:
...
+        structure_factor: Compensated accumulator over complex structure factors,
+            each of shape `(n_groups, n_kvecs, 2)`.
         short_range: Cached real-space short-range output.
         long_range: Cached reciprocal-space long-range output.
         self_interaction: Cached self-interaction correction output.
         exclusion: Cached bonded-pair exclusion correction output.
     """
 
-    structure_factor: Array  # (n_groups, n_kvecs, 2)
+    structure_factor: KahanSummand[Array]  # (n_groups, n_kvecs, 2)
```
Comment (felixmusil) [inline src/kups/potential/classical/ewald.py:116]: I feel like this is the key change from the 3 PRs to enable gcmc loading calculation in fp32. Am I right?

File: test/potential/test_ewald.py [hunk 2]
```diff
@@ -496,3 +503,132 @@ def test_custom_cutoff(self):
...
+            )
+            if not compensated:
+                sk = KahanSummand.init(sk.value)
+            new_pos = pos.at[idx].add(delta)
+            sk, _ = structure_factor(make_input(new_pos, self._cache(sk), previous))
+            return (sk, new_pos), None
+
+        positions = particles.data.positions
+        sk0, _ = structure_factor(make_input(positions))
+        (sk, pos), _ = jax.lax.scan(step, (sk0, positions), (moved, deltas))
```
Comment (felixmusil) [inline test/potential/test_ewald.py:611]: how many `step` are done in this scan?

Comment (n-gao) [reply test/potential/test_ewald.py:611]: `self.N_STEPS` (1000)

File: src/kups/potential/classical/ewald.py (hunk 1 above)
Comment (n-gao) [reply src/kups/potential/classical/ewald.py:116]: The correction on how the self correction term is computed in delta updates it the key enabler

**Review by felixmusil — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Aug 17, 12:00 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/268).
